docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,181 @@
+# ADR 048: Consensus Fees
+
+## Changelog
+
+- Dec 1, 2021: Initial Draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR describes a machanism to maintain the minimal gas price on consensus level.
+
+## Context
+
+Currently, each validator configures it's own `minimal-gas-prices` in `app.yaml`. But setting a proper minimal gas price is critical to protect network from dos attack, and it's hard for all the validators to pick a sensible value, so we proposal to maintain the minimal gas price in consensus level.
+
+Since tendermint 0.35 has supported mempool prioritization, we can take advantage of that to implement more sophisticated gas fee system.
+
+### Compare Different Designs
+
+#### Constant gas fee
+
+##### Positive
+
+- Predictable.
+
+  This makes client behaviour simpler and predictable. And with current system, users are already live with a de facto constant gas price since most validators don't adjust their minimal gas prices.
+
+##### Negative
+
+- Can't include urgent transactions with priority.
+- Don't adjusted based on the block load.
+
+#### First price auction
+
+Before EIP-1559, ethereum miners select transactions ranked by the highest fees, which creates a first price auction market.
+
+##### Positive
+
+- Do provide prioritized service based on tx gas price.
+
+##### Negative
+
+- Unpredictable.
+
+- Overpricing.
+
+#### EIP-1559
+
+[EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) combines a deterministic in-protocol base price and an auctioned tip price.
+
+##### Positive
+
+- Can prioritize urgent transactions.
+- More deterministic than the previous pure auction system.
+
+##### Negative
+
+- Unpredictable.
+- Overpricing the tip price in theory TODO I guess we don't obverse this in real world on ethereum ?
+
+#### Multi-tier pricing system
+
+[Recent IOHK blog](https://iohk.io/en/blog/posts/2021/11/26/network-traffic-and-tiered-pricing/) described a multi-tier gas pricing system. User can specify the tier of service needed for the transaction. The higher the tier, the higher priority, and higher gas price. Within the same tier, transactions pay the same gas price. It avoids the first price auction while provides prioritized service for urgent transactions.
+
+The original proposal reserve block segments for different tiers, we can't do that without modify tendermint, but it should be good enough to just specify higher priority for higher tier transactions. 
+
+The lowest tier can use constant gas price, the higher tiers adjust gas price based on block load at different speed.
```

Comment:
If blocks are consistently full, then won't the lowest tier txs just never make it in, removing the "predictable pricing"?

Reply:
Once we have ABCI++, we reserve a portion of block space for medium and low fee tiers to ensure that this pool of transaction is slowly drained even in full block space environment.

Reply:
Wouldn't that just create an out-of-band tx fee market then?

Reply:
> Wouldn't that just create an out-of-band tx fee market then?

within the same tier, the tx are processed in FIFO manner, avoiding the competition on gas price.